src/main/java/org/apache/maven/buildcache/checksum/MavenProjectInput.java

```diff
@@ -165,45 +184,94 @@ public MavenProjectInput(
+                convertToPathMatcherFileSeperator(
+                                normalizedPath(build.getTestOutputDirectory()).toString())
+                        + GLOB_SX_ALL_SUB_FILES,
+                false); // target/test-classes by default
 
         List<Exclude> excludes = config.getGlobalExcludePaths();
         for (Exclude excludePath : excludes) {
-            filteredOutPaths.add(Paths.get(excludePath.getValue()));
+            addToExcludedSection(excludePath.getValue(), true);
         }
 
         for (String propertyName : properties.stringPropertyNames()) {
             if (propertyName.startsWith(CACHE_EXCLUDE_NAME)) {
                 String propertyValue = properties.getProperty(propertyName);
-                Path path = Paths.get(propertyValue);
-                filteredOutPaths.add(path);
+                addToExcludedSection(propertyValue, true);
+
                 if (LOGGER.isDebugEnabled()) {
                     LOGGER.debug(
-                            "Adding an excludePath from property '{}', values is '{}', path is '{}' ",
-                            propertyName,
-                            propertyValue,
-                            path);
+                            "Adding an excludePath from property '{}', value is '{}'", propertyName, propertyValue);
                 }
             }
         }
         CacheUtils.debugPrintCollection(
-                LOGGER,
-                filteredOutPaths,
-                "List of excluded paths (checked either by fileName or by startsWith prefix)",
-                "Path entry");
+                LOGGER, inputExcludePathMatcherString, "List of excluded glob patterns", "Pattern");
 
         this.fileComparator = new PathIgnoringCaseComparator();
     }
 
+    private String convertToPathMatcherFileSeperator(String path) {
+        return path.replace("\\", "/");
+    }
+
+    /**
+     * Add a value from the excluded section list to the directories and/or the filenames ban list.
+     * @param excludedValue a value from the exclude list
+     */
+    private void addToExcludedSection(String excludedValue, boolean addProjectBaseDir) {
+
+        String pathMatcherGlob = GLOB_PX
+                +
+                // Add the base directory to any input directly coming from user configuration
+                (addProjectBaseDir ? baseDirectoryGlob : "")
+                +
+                // If the glob start with "/", we remove it since it's already added in the added basedir glob
+                (excludedValue.startsWith("/") ? excludedValue.substring(1) : excludedValue);
+
+        // In order to skip unnecessary subtree dir walking, we use a different PathMatcher list for "directories" or
+        // "files + directories"
+        inputExcludePathMatchers.add(new TreeWalkerPathMatcher(pathMatcherGlob, false));
+
+        // Globs ending with "**" should end any sub-directory inspection
+        if (pathMatcherGlob.endsWith("**")) {
```

Comment:
I totally agree, even if the flexibility offered is pretty nice, I wasn't so pleased with the complexity of this demand.
As a trade-off, my strategy was : 
- if the syntax is not written in an exotic manner : exclusion is fast as it can stop tree walking
- if not : exclusion is slower but works as expected

Reply:
Well, there are different semantics now and before for project-level properties.
Before: declaring `mydir/mysubdir` exclusion was filtering out the whole subtree (through the `Path#startsWith`)
Now: declaring `mydir/mysubdir` doesn't exclude anything.
```
FileSystems.getDefault().getPathMatcher("glob:dir");
System.out.println(pathMatcher.matches(Paths.get("dir/file.txt")));`
-----OUT-----
false
```
 For consistency, we must keep the path semantics of properties or migrate them to the globs/regex semantics, delegating filtering expressions to the users. But the main point is that it should be consistent and simple.